In [10]:
import pandas as pd

# Load the CSV file into a DataFrame
df = pd.read_csv('/users/andrewreis/desktop/villagers.csv')

# Display the first few rows of the DataFrame
df.head()

row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig     4-21        uchi   
3      6       al       Al    male    gorilla    10-18        lazy   
4      7  alfonso  Alfonso    male  alligator      6-9        lazy   

          song    phrase           full_id  \
0   Steep Hill   aye aye  villager-admiral   
1      DJ K.K.  sidekick  villager-agent-s   
2   K.K. House   snuffle    villager-agnes   
3   Steep Hill   Ayyeeee       villager-al   
4  Forest Life  it'sa me  villager-alfonso   

                                                 url  
0  https://villagerdb.com/images/villagers/thumb/...  
1  https://villagerdb.com/images/villagers/thumb/...  
2  https://villagerdb.com/images/villagers/thumb/...  
3  https://villagerdb.com/images/villagers/thumb/...  
4  https://villagerdb.com/images/villagers/thumb/...

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


In [14]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [18]:
df['species'].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

In [20]:
df['gender'].value_counts()

gender
male      204
female    187
Name: count, dtype: int64

In [22]:
df['birthday'].value_counts()

birthday
1-27     2
12-5     2
7-31     2
3-26     2
8-3      2
        ..
4-3      1
10-26    1
7-23     1
12-8     1
3-8      1
Name: count, Length: 361, dtype: int64

In [24]:
df['personality'].value_counts()

personality
lazy      60
normal    59
cranky    55
snooty    55
jock      55
peppy     49
smug      34
uchi      24
Name: count, dtype: int64

In [28]:
df_titanic = pd.read_csv('/users/andrewreis/desktop/titanic.csv')

In [30]:
df_titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [32]:
df_titanic.shape

(891, 15)

In [34]:
print("Shape of the dataset:", df_titanic.shape)

Shape of the dataset: (891, 15)


In [36]:
print("Descriptive statistics:")
print(df_titanic.describe())

Descriptive statistics:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


In [38]:
print("Missing values in each column:")
print(df_titanic.isnull().sum())

Missing values in each column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


In [48]:
print("Value Counts for 'gender':")

Value Counts for 'gender':


### Chat Summary and pre Lecture HW answers

- **Dataset Analysis**: We used `df.describe()` to get statistical summaries of numeric columns and `df['column'].value_counts()` to understand the distribution of non-numeric columns.
- **Handling Missing Values**: We examined the differences between `df.shape` (the size of the dataset) and `df.describe()` (which only shows numeric columns and may have discrepancies if there are missing values).
- **Attributes vs. Methods**: Attributes (e.g., `df.shape`) provide information about an object without needing to be executed, while methods (e.g., `df.describe()`) perform actions and require parentheses to execute.

Question 2 Answer: Observations: The Titanic for example has rows, and each of these rows consists of a single passengers data, like age, survived or not, class, and more.                                                                                                                 
                   Variables: The variables are the collumns in this dataset, those include the age, class, gender. Except they give the data for every passengers, unlike the rows which only have data for a specific passenger.                                                                                                 
Question 4 Answer: For the titanic dataset, we can see that the almost all the missing values are the age. I also see that there are 891 rows and only 15 collumns.
Something I noticed is that there was definetly more than 900 people on the boat, but it makes sense they couldnt get all the data, and if they did, im sure there would be alot more missing values.
Using help from ChatGPT, the difference between df.shape and df.describe() is that df.describe() excudes non numerical collumns.                                                                            
Question 5 Answer: My understanding of attributes vs methods is that methods is performing a task and attributes is just showing us the informations. Also it is important to note that methods has brackets at the end of it, making it a function.                                                                                                                 
                                                                                                                     
                                                                                                                     
For more detailed information, refer to the chat log history [here](https://chatgpt.com/share/66e446b8-18d0-8012-b4ce-3345914ba954).

In [50]:
df['sex'].value_counts()

KeyError: 'sex'

### Post lecture HW

6: Definitions
Count: The count is the total number of values or data that is not missing from the data set.
Mean: Mean is the average of every single number in a collumn. To find it, you must add up all the numbers in the collumn and divide it by the count.
Standard Deviation: Standard deviation shows you how much the data is spread out from the average. Having a large standard deviation means the data is very spread out.
Having a small one means the data is not very spread out.
min: the smallest number in the collumn.
max: the biggest number in the collumn.
25%: Described as "the first checkpoint" 25% of the values are below it.
50%: This is the middle number of the data set, half of the numbers are below it and half the numbers are above
75%:the third checkpoint in the data, 75% of the values are below this number


7: Use cases for df.dropna() 
df.dropna() is can be used to drop certain rows that are missing some key elements of data. For example, with the titanic data, we are missing some stats that are related to the age of the people on the ship. If were solely trying to focus on using the data of those with a represented age, this code would be good to drop those that have none.

Use cases for del df['col'] 
del df['col'] can be used for columns with missing values. If there were alot more missing values in the age column (say around 90% were missing) it would probably be best to just drop it completely.

When should you use del df['col'] and when should you use df.dropna(). It depends on the scenario and the amount of missing values. If you have 90% missing values of a certain collumn, its best to use df['col']. You'll save yourself a lot more time if you do. If you have a couple rows where theres multiple missing values, you should use df.dropna(). You should use df['col'] first if your planning on using both. It will eliminate most missing values, and then you can figure out which rows still have missing values and remove those afterwards. That would be best if your trying to save time, which is a huge thing when coding.

In [63]:
# Load the Titanic dataset
df_titanic = pd.read_csv('/users/andrewreis/desktop/titanic.csv')

# Display the first few rows of the dataset
df_titanic.head()# Check for missing values in each column
missing_data = df_titanic.isna().sum()
print("Missing values in each column before cleaning:")
print(missing_data)

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [65]:
# Check for missing values in each column
missing_data = df_titanic.isna().sum()
print("Missing values in each column before cleaning:")
print(missing_data)

Missing values in each column before cleaning:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


In [67]:
# Remove the 'deck' column
del df_titanic['deck']

In [69]:
# Drop rows with missing values in 'age', 'embarked', and 'embark_town'
df_titanic_cleaned = df_titanic.dropna(subset=['age', 'embarked', 'embark_town'])

In [71]:
# Check for remaining missing values
print("Missing values after cleaning:")
print(df_titanic_cleaned.isna().sum())

Missing values after cleaning:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64


7: I removed the deck collumn completely because it had way too many missing values to be helpful with the over all data. Then afterwards I removed the rows that were missing the age, embark_town, and embarked values. Now there are no missing values whatsoever and our total data set feels a lot more complete.

Here's a summary of our chats so far starting from the "Post-lecture" homework:

Prompt 6 - Summary Statistics Definitions (df.describe()):
We discussed the meanings of the summary statistics provided by df.describe() like count, mean, std, min, max, and percentiles (25%, 50%, 75%). You requested a simplified explanation, which I provided, focusing on their basic roles and how they are calculated only for numeric columns.
Prompt 7 - Handling Missing Data (df.dropna() vs del df['col']):
We worked through how to efficiently remove missing data by using df.dropna() to remove rows with missing values and del df['col'] to remove columns with excessive missing data. We applied this to the Titanic dataset, where the deck column had too many missing values, and rows with missing values in age, embarked, and embark_town were dropped.
Prompt 8 - Troubleshooting Code with Errors:
We covered how to troubleshoot common coding errors such as forgetting to import pandas (NameError), mistyping filenames or column names (FileNotFoundError, KeyError), or introducing syntax errors. We discussed whether it’s more efficient to use a ChatBot or Google for fixing such issues.
This summary highlights the main tasks we’ve worked through, with a focus on cleaning data and troubleshooting in the Titanic dataset. Let me know if you'd like any further details! https://chatgpt.com/share/66e446b8-18d0-8012-b4ce-3345914ba954

In [75]:
df.groupby("age")["Survived"].describe()

KeyError: 'age'

In [77]:
df.groupby(df.columns[3])["Survived"].describe()  # Adjust index if "age" is at a different position

KeyError: 'Column not found: Survived'

In [79]:
df.groupby(df.columns[3])["survived"].describe()  # Adjust index if "age" is at a different position

KeyError: 'Column not found: survived'

In [81]:
print(df.columns)

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')


In [83]:
import pandas as pd

df = pd.read_csv('path/to/titanic.csv') 

FileNotFoundError: [Errno 2] No such file or directory: 'path/to/titanic.csv'

In [85]:
import pandas as pd

df = pd.read_csv('/users/andrewreis/desktop/titanic.csv') 

In [87]:
print(df.columns)

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


In [89]:
df.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

8: It looks like there is a HUGE gap between the survival rates of the people on board based on gender. a 75% survival rate for women and a 19% survival rate for men.

Question 8.3: It is 100x easier to use the chatbot for fixing errors than google and it isnt even close. ChatGPT is instantly giving me answers that are specific to my problem and their solutions have not failed me yet. I honestly really think that ChatGPT is the most amazing tool for getting this work done efficiently. The only problem is its easy to copy and paste but it isnt exactly easy to learn the actual coding. I kind of understanding whats happening, but if I was strictly doing this on my own I would probably take longer, but I'd actually learn it and understand it better in the long run.

In [1]:
df_titanic = pd.read_csv('/users/andrewreis/desktop/titanic.csv')

NameError: name 'pd' is not defined

In [3]:
import pandas as pd

# Load the Titanic dataset
df_titanic = pd.read_csv('/users/andrewreis/desktop/titanic.csv')

# Verify that the data is loaded
print(df_titanic.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [5]:
df_titanic = pd.read_csv('/users/andrewreis/desktop/titanics.csv')

FileNotFoundError: [Errno 2] No such file or directory: '/users/andrewreis/desktop/titanics.csv'

In [7]:
DF.groupby("col1")["col2"].describe()


NameError: name 'DF' is not defined

In [9]:
pd.read_csv(url

SyntaxError: incomplete input (4098100527.py, line 1)

In [11]:
df.groupby("sex")["survived"].describle()

NameError: name 'df' is not defined

In [13]:
titanic_df.groupby("sex")["age"].describe()

NameError: name 'titanic_df' is not defined

In [15]:
import pandas as pd

# Replace with the correct path to your CSV file
df = pd.read_csv('/users/andrewreis/desktop/titanic.csv')

In [17]:
df.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

In [19]:
titanic_df.groupby("sex")["age"].describe()

NameError: name 'titanic_df' is not defined

In [21]:
import pandas as pd

# Load the Titanic dataset
df = pd.read_csv('/users/andrewreis/desktop/titanic.csv')  # Ensure this path is correct

In [23]:
df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

In [25]:
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

In [27]:
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

Another thing that is amazing about ChatBots for coding is that it will return to you the next line(s) of code that you should be using. It already has them pre typed for you. Google may tell you how to fix it, but it is not nearly as efficient.

A: After not putting the code import pandas as pd, i put the error into chatGPT and it instantly told me what was wrong very simply, it also gave me the exact line of code, and some other lines of code to fix it and test. This is way better than googles response since the explanations and code given back to me were much more clear.

B: Spelling the file “titanics.csv” gives me an error that when pasting into chatGPT gives me multiple reasons for why this occurred. This was very helpful and it did notice the problem, stating that I should double check if it was spelt correctly or not.

C: Once again, chat gpt recognizes that there is an error, it tells me that it believes I meant to spell “df” instead of “DF” which is completely true. Another correct solution from chatgpt.

D: ChatGPT recognizes the error, stating that i need to add a closing bracket on the end of my code to remove the error. It then gives me a couple new lines of code so that I can copy and paste into my jupyter notebook and test if their solution was correct. Which it was.

E: Chat gpts response to my typo of “describl” was that of course, I had a typo and that needed to be fixed. It then responds by giving me a new line of code, with everything spelt correctly, and it already gives me the next couple lines of code that I may want to use.

F: when putting in “Sex” instead of “sex” it gives me an error that when pasted into chat gpt, chat gpt tells me its possible that the column im talking about doesnt exist, it then tells me that I should use another command to see all the names of the columns, a very useful tip. It also tells me that if i’m not using the exact name, it won’t work, which turns out to be the reason for the error.

G: For the next error, we dont put the word sex in quotes, which causes an error. When pasting this error into gpt, it gives us multiple reasons for why this is happening, but it does gives us the correct solution, telling us to put it in quotes. Followed up by that correction, it then gives us the correct line of code.


Here’s a summary of what we’ve done since the NameError related to pd:

Error Encountered:
Issue: NameError: name 'pd' is not defined
Cause: The pandas library was not imported.
Solution:
Action: Imported pandas using import pandas as pd and attempted to load the Titanic dataset.
FileNotFoundError:
Issue: FileNotFoundError indicating the file titanics.csv could not be found.
Cause: Incorrect file path or file name.
Verification Steps:
Action: Suggested checking the file path and file name, and ensuring the correct path and name were used.
Kernel Restart:
Issue: Kernel restart might have cleared all previous definitions.
Action: Re-loaded the DataFrame and verified its content.
KeyError:
Issue: KeyError: 'Sex'
Cause: Column name 'Sex' did not exist or was misspelled.
Action: Suggested checking the actual column names and updating code to use the correct names.
NameError with Variable:
Issue: NameError: name 'titanic_df' is not defined
Cause: Variable titanic_df was not defined or used incorrectly.
Action: Re-defined the DataFrame variable correctly and re-loaded the dataset.
Syntax Error in pd.read_csv:
Issue: SyntaxError: incomplete input
Cause: Incomplete or incorrect syntax in pd.read_csv.
Action: Provided correct syntax and ensured proper usage of parentheses.
Re-checking DataFrame:
Action: Printed column names and confirmed the correct columns to use.
Final Errors and Fixes:
NameError: NameError: name 'sex' is not defined
Cause: Used column name without quotes.
Action: Ensured column names were correctly enclosed in quotes.
GroupBy and Describe:
Action: Attempted to group data by columns and describe statistics, ensuring column names were correct.
